pkg/trace/api/normalizer_test.go

```diff
@@ -33,277 +36,326 @@ func testSpan() *pb.Span {
 	}
 }
 
+// assertNormalizationIssue asserts there is exactly one NormalizationIssue in the provided TagStats matching the reason
+func assertNormalizationIssue(t *testing.T, ts *info.TagStats, reason string) {
+	allIssues := ts.AllNormalizationIssues()
+
+	presentIssues := make(map[string]int64)
+	for r, c := range allIssues {
+		if c > 0 {
+			presentIssues[r] = c
+		}
+	}
+
+	foundIssuesMessage := "none"
+	if len(presentIssues) > 0 {
+		s, _ := json.Marshal(presentIssues)
+		foundIssuesMessage = string(s)
+	}
+
+	if count, ok := presentIssues[reason]; ok {
+		assert.EqualValues(t, 1, count, fmt.Sprintf("expected normalization issue %s=%d. Found issues: %s", reason, 1, foundIssuesMessage))
+	} else {
+		assert.Fail(t, fmt.Sprintf("missing expected normalization issue %s=%d. Found issues: %s", reason, 1, foundIssuesMessage))
+	}
+}
+
+// assertNormalizationIssue asserts there there are no NormalizationIssues in the provided TagStats
+func assertNoNormalizationIssues(t *testing.T, ts *info.TagStats) {
+	for reason, count := range ts.AllNormalizationIssues() {
+		assert.EqualValues(t, 0, count, fmt.Sprintf("found unexpected normalization issue %s=%d, expecting no issues", reason, count))
+	}
+}
+
 func TestNormalizeOK(t *testing.T) {
-	s := testSpan()
-	assert.NoError(t, normalize(s))
+	ts, s := &info.TagStats{}, newTestSpan()
+	assert.NoError(t, normalize(ts, s))
+	assertNoNormalizationIssues(t, ts)
 }
 
 func TestNormalizeServicePassThru(t *testing.T) {
-	s := testSpan()
+	ts, s := &info.TagStats{}, newTestSpan()
 	before := s.Service
-	normalize(s)
+	assert.NoError(t, normalize(ts, s))
 	assert.Equal(t, before, s.Service)
+	assertNoNormalizationIssues(t, ts)
+}
+
+func TestNormalizeEmptyServiceNoLang(t *testing.T) {
+	ts, s := &info.TagStats{}, newTestSpan()
+	s.Service = ""
+	assert.NoError(t, normalize(ts, s))
+	assert.Equal(t, s.Service, DefaultServiceName)
+	assertNormalizationIssue(t, ts, "service_empty")
```

Comment:
That's cool, didn't realize the tests could compare structs in detail. 

I think the current approach is better for several reasons:

* it's testing the exact reason names `service_empty` used in tags, logs, and agent status instead of the internal counters `TracesMalformed.ServiceEmpty`. It's better to test that because we're testing the "interface" and not the underlying implementation. If we were testing only the counters we might not catch that a tag name was wrong. 
* `assertNormalizationIssue(t, ts, "service_empty")` is a bit easier to read than `assert.Equal(t, ts, tagStatsMalformed(&info.TracesMalformed{ServiceEmpty: 1})`
* broken test output is a bit clearer since we're only comparing the present issues and it won't show a lot of other unrelated spam like all the zero values when it does `%#v` formatting of the entire structs

But I like the idea of comparing objects directly. Updated to use direct comparison of maps instead of hacky manual diff https://github.com/DataDog/datadog-agent/pull/3738/commits/86d706ec95d3df9f5f833deb1a1a2eee8080ea68 what do you think of that?

Reply:
> `assertNormalizationIssue(t, ts, "service_empty")` is a bit easier to read than `assert.Equal(t, ts, tagStatsMalformed(&info.TracesMalformed{ServiceEmpty: 1})`

I disagree. Nobody knows what `assertNormalizationIssue` unless they go to a definition and learn a new function. The latter is already known.

> broken test output is a bit clearer since we're only comparing the present issues and it won't show a lot of other unrelated spam like all the zero values when it does %#v formatting of the entire structs

Have you tried it? The output is pretty good.

Reply:
> Have you tried it? The output is pretty good.

You're right,  the diff output is actually fine, so it's good to use the object comparison. 

I updated to have all asserts done right in the tests, not in these other functions. Still using a bit of code to simplify the tests, but I think that's fairly reasonable. 

It now looks like this:

Asserting no issues: `assert.Equal(t, noIssues(), statsToIssues(ts))`
Asserting a specific issue: `assert.Equal(t, onlyIssue("span_name_empty"), statsToIssues(ts))`

I think this should be ok?